TrueCraft.Core.Test/World/SugarCaneDecoratorTests.cs

```diff
@@ -0,0 +1,461 @@
+                }
+            }
+            return counter;
+        }
+
+
+        public static HashSet<Coordinates2D> PointsWithoutAnySugarcane()
+        {
+            var result = new HashSet<Coordinates2D>();
+            var startingDictionary = createStartingBlockDictionary();
+            foreach (var block in createStartingBlockDictionary())
+            {
+                var placesToGrow = CountBlockInColumn(startingDictionary, block.Key.X, block.Key.Z, SandBlock.BlockID);
+                placesToGrow += CountBlockInColumn(startingDictionary, block.Key.X, block.Key.Z, GrassBlock.BlockID);
+
+                if (placesToGrow == 0)
+                {
+                    result.Add(new Coordinates2D(block.Key.X, block.Key.Z));
+                }
+            }
+
+            result.Add(new Coordinates2D(0, 0));
+            result.Add(new Coordinates2D(0, 1));
+            result.Add(new Coordinates2D(0, 0));
+            result.Add(new Coordinates2D(0, 2));
+            result.Add(new Coordinates2D(0, 4));
+            result.Add(new Coordinates2D(0, 5));
+
+
+            for (int i = 1; i < 5; i++)
+            {
+                // Top row of the test data is all landlocked
+                result.Add(new Coordinates2D(i, 5));
+            }
+
+            result.Add(new Coordinates2D(5, 5));
+            result.Add(new Coordinates2D(5, 4));
+            result.Add(new Coordinates2D(5, 2));
+            result.Add(new Coordinates2D(5, 1));
+
+            return result;
+        }
+    }
+
+
+    [TestFixture]
+    public class SugarCaneDecoratorTests
+    {
+        bool ContainsSugarCane(IChunk aChunk, int x, int z)
+        {
+            return CountBlockInColumn(aChunk, x, z, SugarcaneBlock.BlockID) > 0;
+        }
+        
+        static int CountBlockInColumn(IChunk aChunk, int x, int z, byte blockId)
+        {
+            int counter = 0;
+
+            for (int y = 0; y < 7; y++)
+            {
+                byte block = aChunk.GetBlockID(new Coordinates3D(x:x, y:y, z:z));
+                if (block == blockId)
+                {
+                    counter++;
+                }
+            }
+            return counter;
+        }
+
+        
+        [Test]
+        public void TestBasicDecorator()
+        {
+            var decorator = new SugarCaneDecorator(new NoiseAlwaysGrowsSugarCaneInTestBounds());
+            IWorld aWorld = new WorldWithJustASeed(9001);
+            IChunk aChunk = new PrimeSugarCaneGrowingSeasonChunk();
+            IBiomeRepository aBiomeRepository = new BiomeRepository();
+            aBiomeRepository.RegisterBiomeProvider(new SwamplandBiome());
+            
+            decorator.Decorate(aWorld, aChunk, aBiomeRepository);
+
+            AssertChunkHasNoSugarCaneInColumnsWhereItShouldNot(aChunk);
+            AssertChunkSugarCaneGrowthIsNotUniform(aChunk);
+        }
+
+        private void AssertChunkHasNoSugarCaneInColumnsWhereItShouldNot(IChunk aChunk)
```

Comment:
These should be separate tests, and probably with better names.

NoInvalidGrowth

NotUniform

Reply:
👍 agreed. Fixing